File: skills/algotrader/trading_config.py
from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path
from typing import Any
import tomllib


_CONFIG_DIR = Path(__file__).resolve().parent / "config"
_DEFAULTS_PATH = _CONFIG_DIR / "trading.defaults.toml"
_LOCAL_PATH = _CONFIG_DIR / "trading.local.toml"
_PLACEHOLDER = "__SET_IN_trading.local.toml__"


def _merge_dicts(base: dict[str, Any], override: dict[str, Any]) -> dict[str, Any]:
    merged = dict(base)
    for key, value in override.items():
        if isinstance(value, dict) and isinstance(merged.get(key), dict):
            merged[key] = _merge_dicts(merged[key], value)
        else:
            merged[key] = value
    return merged


def _load_toml(path: Path) -> dict[str, Any]:
    if not path.exists():
        return {}
    with path.open("rb") as fh:
        return tomllib.load(fh)


def _coerce_path(value: str, field: str) -> Path:
    if not value or value == _PLACEHOLDER:
        raise RuntimeError(f"Missing required algotrader config path: {field}")
    return Path(value).expanduser().resolve()


@dataclass(frozen=True)
class Profile:
    name: str
    dataset: str
    timeframe: str
    venue: str | None
    bar_spec: str | None
    universe: Path | None
    raw_dir: Path | None
    ohlcv: Path | None
    catalog: Path | None


@dataclass(frozen=True)
class TradingConfig:
    repo_root: Path
    data_root: Path
    nautilus_repo_root: Path
    lab_notebook_path: Path
    strategies_dir: Path
    engine_binary: Path
    datasets: dict[str, Path]
    artifacts: dict[str, Path]
    profiles: dict[str, Profile]
    integration: dict[str, str]

    def profile(self, name: str) -> Profile:
        try:
            return self.profiles[name]
        except KeyError as exc:
            known = ", ".join(sorted(self.profiles))
            raise RuntimeError(f"Unknown algotrader profile '{name}'. Known profiles: {known}") from exc


@lru_cache(maxsize=1)
def load_trading_config() -> TradingConfig:
    if not _LOCAL_PATH.exists():
        raise RuntimeError(
            f"Missing local algotrader config at {_LOCAL_PATH}. "
            f"Copy {_DEFAULTS_PATH.name} to {_LOCAL_PATH.name} and fill in real paths."
        )

    merged = _merge_dicts(_load_toml(_DEFAULTS_PATH), _load_toml(_LOCAL_PATH))
    paths = merged.get("paths", {})
    datasets = merged.get("datasets", {})
    artifacts = merged.get("artifacts", {})
    profiles_raw = merged.get("profiles", {})
    integration = merged.get("integration", {})

    resolved_datasets = {
        key: _coerce_path(value, f"datasets.{key}")
        for key, value in datasets.items()
    }
    resolved_artifacts = {
        key: _coerce_path(value, f"artifacts.{key}")
        for key, value in artifacts.items()
    }

    profiles: dict[str, Profile] = {}
    for name, profile in profiles_raw.items():
        profiles[name] = Profile(
            name=name,
            dataset=profile["dataset"],
            timeframe=profile["timeframe"],
            venue=profile.get("venue"),
            bar_spec=profile.get("bar_spec"),
            universe=_coerce_path(profile["universe"], f"profiles.{name}.universe") if profile.get("universe") else None,
            raw_dir=_coerce_path(profile["raw_dir"], f"profiles.{name}.raw_dir") if profile.get("raw_dir") else None,
            ohlcv=_coerce_path(profile["ohlcv"], f"profiles.{name}.ohlcv") if profile.get("ohlcv") else None,
            catalog=_coerce_path(profile["catalog"], f"profiles.{name}.catalog") if profile.get("catalog") else None,
        )

    return TradingConfig(
        repo_root=_coerce_path(paths["repo_root"], "paths.repo_root"),
        data_root=_coerce_path(paths["data_root"], "paths.data_root"),
        nautilus_repo_root=_coerce_path(paths["nautilus_repo_root"], "paths.nautilus_repo_root"),
        lab_notebook_path=_coerce_path(paths["lab_notebook_path"], "paths.lab_notebook_path"),
        strategies_dir=_coerce_path(paths["strategies_dir"], "paths.strategies_dir"),
        engine_binary=_coerce_path(paths["engine_binary"], "paths.engine_binary"),
        datasets=resolved_datasets,
        artifacts=resolved_artifacts,
        profiles=profiles,
        integration=dict(integration),
    )

File: skills/algotrader/engine/src/trading_config.rs
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::OnceLock;

use anyhow::{Context, Result};
use serde::Deserialize;
use toml::Value;


static TRADING_CONFIG: OnceLock<TradingConfig> = OnceLock::new();
const PLACEHOLDER: &str = "__SET_IN_trading.local.toml__";

#[derive(Clone, Debug, Deserialize)]
pub struct TradingConfig {
    pub paths: PathsConfig,
    pub datasets: HashMap<String, String>,
    pub artifacts: HashMap<String, String>,
    pub profiles: HashMap<String, ProfileConfig>,
    pub integration: HashMap<String, String>,
}

#[derive(Clone, Debug, Deserialize)]
pub struct PathsConfig {
    pub repo_root: String,
    pub data_root: String,
    pub nautilus_repo_root: String,
    pub lab_notebook_path: String,
    pub strategies_dir: String,
    pub engine_binary: String,
}

#[derive(Clone, Debug, Deserialize)]
pub struct ProfileConfig {
    pub dataset: String,
    pub timeframe: String,
    pub venue: Option<String>,
    pub bar_spec: Option<String>,
    pub universe: Option<String>,
    pub raw_dir: Option<String>,
    pub ohlcv: Option<String>,
    pub catalog: Option<String>,
}

fn merge_toml(base: &mut Value, overlay: Value) {
    match (base, overlay) {
        (Value::Table(base_map), Value::Table(overlay_map)) => {
            for (key, value) in overlay_map {
                match base_map.get_mut(&key) {
                    Some(existing) => merge_toml(existing, value),
                    None => {
                        base_map.insert(key, value);
                    }
                }
            }
        }
        (slot, value) => *slot = value,
    }
}

fn read_toml(path: &Path) -> Result<Value> {
    let content = fs::read_to_string(path)
        .with_context(|| format!("Failed to read TOML config {}", path.display()))?;
    let value: Value = toml::from_str(&content)
        .with_context(|| format!("Failed to parse TOML config {}", path.display()))?;
    Ok(value)
}

fn require_real_path(value: &str, field: &str) -> Result<()> {
    if value.is_empty() || value == PLACEHOLDER {
        anyhow::bail!("Missing required algotrader config value: {field}");
    }
    Ok(())
}

fn validate(cfg: &TradingConfig) -> Result<()> {
    require_real_path(&cfg.paths.repo_root, "paths.repo_root")?;
    require_real_path(&cfg.paths.data_root, "paths.data_root")?;
    require_real_path(&cfg.paths.nautilus_repo_root, "paths.nautilus_repo_root")?;
    require_real_path(&cfg.paths.lab_notebook_path, "paths.lab_notebook_path")?;
    require_real_path(&cfg.paths.strategies_dir, "paths.strategies_dir")?;
    require_real_path(&cfg.paths.engine_binary, "paths.engine_binary")?;
    for (key, value) in &cfg.datasets {
        require_real_path(value, &format!("datasets.{key}"))?;
    }
    for (key, value) in &cfg.artifacts {
        require_real_path(value, &format!("artifacts.{key}"))?;
    }
    for (name, profile) in &cfg.profiles {
        if !cfg.datasets.contains_key(&profile.dataset) {
            anyhow::bail!(
                "Profile '{name}' references unknown dataset '{}'",
                profile.dataset
            );
        }
        if let Some(value) = &profile.ohlcv {
            require_real_path(value, &format!("profiles.{name}.ohlcv"))?;
        }
    }
    Ok(())
}

pub fn default_config_paths() -> (PathBuf, PathBuf) {
    let repo_root = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("engine has parent directory")
        .to_path_buf();
    (
        repo_root.join("config").join("trading.defaults.toml"),
        repo_root.join("config").join("trading.local.toml"),
    )
}

pub fn load_trading_config() -> Result<TradingConfig> {
    let (defaults_path, local_path) = default_config_paths();
    let mut merged = read_toml(&defaults_path)?;
    if !local_path.exists() {
        anyhow::bail!(
            "Missing local algotrader config at {}",
            local_path.display()
        );
    }
    merge_toml(&mut merged, read_toml(&local_path)?);
    let cfg: TradingConfig = merged
        .try_into()
        .context("Failed to deserialize algotrader trading config")?;
    validate(&cfg)?;
    Ok(cfg)
}

pub fn set_trading_config(cfg: TradingConfig) {
    let _ = TRADING_CONFIG.set(cfg);
}

pub fn trading_config() -> Option<&'static TradingConfig> {
    TRADING_CONFIG.get()
}

pub fn dataset_dir_for_profile(profile: &str) -> Result<PathBuf> {
    let cfg = trading_config().context("Trading config has not been initialized")?;
    let profile_cfg = cfg
        .profiles
        .get(profile)
        .with_context(|| format!("Unknown algotrader profile '{profile}'"))?;
    let dataset = cfg
        .datasets
        .get(&profile_cfg.dataset)
        .with_context(|| format!("Profile '{profile}' points at unknown dataset '{}'", profile_cfg.dataset))?;
    Ok(PathBuf::from(dataset))
}

pub fn strategies_dir() -> Option<PathBuf> {
    trading_config().map(|cfg| PathBuf::from(&cfg.paths.strategies_dir))
}
